Add SupportedOSPlatform to Hosting.WindowsService HostBuilder extensions

hwoodiwiss opened the pull request:
fixes dotnet/runtime#50013

Includes System.Runtime.Versioning attributes into the build and applies them to the HostBuilder extensions.

hwoodiwiss committed:
Add build property to expose attibute

Added IncludePlatformAttributes build property to allow
SupportedOSPlatform to be exposed to the assembly.

hwoodiwiss committed:
Add Supported OS Attribute to Extensions

Add SupportedOSPlatform attribute to UseWindowsService extension methods

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53941) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53941) before we can accept your contribution.<br/><br/>:x: hwoodiwiss [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53941)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=53941) it.</sub>

hwoodiwiss commented:
Feedback is appreciated, I was on the fence between applying the attribute to the extension methods, and just applying it as an assembly attribute.

hwoodiwiss committed:
Merge branch 'dotnet:main' into hwoodiwiss-50013

hwoodiwiss committed:
Add the SupportedOSPlatform attribute to refs

Add the SupportedOSPlatform attribute to corresponding methods in refs.

hwoodiwiss committed:
Correct SupportedOSPlatform placement in refs

Update refs for Microsoft.Extensions.Hosting.WindowsServices
to correct incorrect placement of SupportedOSPlatform attribute

eerhardt commented:
> I was on the fence between applying the attribute to the extension methods, and just applying it as an assembly attribute.

I think it makes the most sense to be assembly wide. The whole assembly is "WindowsServices". Those are only ever going to be supported on Windows.

We do it assembly wide elsewhere:

https://github.com/dotnet/runtime/blob/197cfb60f37626ce9e07285517b129fc01f9621e/src/libraries/Microsoft.Win32.Registry/Directory.Build.props#L6
https://github.com/dotnet/runtime/blob/197cfb60f37626ce9e07285517b129fc01f9621e/src/libraries/System.Security.Principal.Windows/Directory.Build.props#L6

cc @jeffhandley @buyaa-n

hwoodiwiss committed:
Revert attribute changes to UseWindowsService

Revert addition of SupportedOSPlatform attribute to UseWindowsService
extension methods and refs for them

hwoodiwiss committed:
Set SupportedOSPlatforms at assembly level

Update build props to set SupportedOSPlatforms to windows assembly wide

hwoodiwiss commented:
Great feedback and great point. I've update to set the attribute assembly wide as suggested.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetimeHostBuilderExtensions.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
```
Is this change still necessary?

hwoodiwiss replied:
It was not, good catch, sorry about that.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/ref/Microsoft.Extensions.Hosting.WindowsServices.cs:
```diff
@@ -8,8 +8,8 @@ namespace Microsoft.Extensions.Hosting
     {
-        public static Microsoft.Extensions.Hosting.IHostBuilder UseWindowsService(this Microsoft.Extensions.Hosting.IHostBuilder hostBuilder) { throw null; }
-        public static Microsoft.Extensions.Hosting.IHostBuilder UseWindowsService(this Microsoft.Extensions.Hosting.IHostBuilder hostBuilder, System.Action<Microsoft.Extensions.Hosting.WindowsServiceLifetimeOptions> configure) { throw null; }
+       public static Microsoft.Extensions.Hosting.IHostBuilder UseWindowsService(this Microsoft.Extensions.Hosting.IHostBuilder hostBuilder) { throw null; }
```
I believe this change should be reverted.

hwoodiwiss replied:
This has now been reverted.

hwoodiwiss committed:
Complete reverting unnecesarry changes

Add back removed whitespace in ref
Remove unused using of System.Runtime.Versioning

eerhardt reviewed (approved):
Looks great. Thanks for the contribution!
